api/src/test/java/org/pmiops/workbench/notebooks/NotebooksServiceTest.java

```diff
@@ -98,39 +133,416 @@ public void setup() {
+    // and thatg is tested below.
+    notebooksService.cloneNotebook(NAMESPACE_NAME, WORKSPACE_NAME, PREVIOUS_NOTEBOOK);
+
+    verify(mockWorkspaceAuthService)
+        .enforceWorkspaceAccessLevel(NAMESPACE_NAME, WORKSPACE_NAME, WorkspaceAccessLevel.READER);
+    verify(mockWorkspaceAuthService)
+        .enforceWorkspaceAccessLevel(NAMESPACE_NAME, WORKSPACE_NAME, WorkspaceAccessLevel.WRITER);
+    verify(mockWorkspaceAuthService).validateActiveBilling(NAMESPACE_NAME, WORKSPACE_NAME);
+    verify(mockLogsBasedMetricsService).recordEvent(EventMetric.NOTEBOOK_CLONE);
+  }
+
+  @Test
+  public void testCopyNotebook() {
+    String fromWorkspaceNamespace = "fromWorkspaceNamespace";
+    String fromWorkspaceFirecloudName = "fromWorkspaceFirecloudName";
+    String fromNotebookName = "fromNotebookName";
+    String toWorkspaceNamespace = "toWorkspaceNamespace";
+    String toWorkspaceFirecloudName = "toWorkspaceFirecloudName";
+    String newNotebookName = "newNotebookName";
+    String bucketName = "the_bucket";
+    HashSet<BlobId> existingBlobIds = new HashSet<>();
+
+    fromCDRVersion.setAccessTier(fromAccessTier);
+    toCDRVersion.setAccessTier(toAccessTier);
+
+    stubGetWorkspace(
+        fromWorkspaceNamespace, fromWorkspaceFirecloudName, WorkspaceAccessLevel.OWNER);
+    stubGetWorkspace(toWorkspaceNamespace, toWorkspaceFirecloudName, WorkspaceAccessLevel.OWNER);
+
+    doReturn(dbWorkspace).when(workspaceDao).getRequired(anyString(), anyString());
+    when(mockWorkspaceAuthService.enforceWorkspaceAccessLevel(anyString(), anyString(), any()))
+        .thenReturn(WorkspaceAccessLevel.OWNER);
+
+    when(workspaceDao.getRequired(fromWorkspaceNamespace, fromWorkspaceFirecloudName))
+        .thenReturn(dbWorkspace);
+    when(workspaceDao.getRequired(toWorkspaceNamespace, toWorkspaceFirecloudName))
+        .thenReturn(dbWorkspace);
+
+    when(mockCloudStorageClient.getExistingBlobIdsIn(any())).thenReturn(existingBlobIds);
+
+    FileDetail actualFileDetail =
+        notebooksService.copyNotebook(
+            fromWorkspaceNamespace,
+            fromWorkspaceFirecloudName,
+            fromNotebookName,
+            toWorkspaceNamespace,
+            toWorkspaceFirecloudName,
+            newNotebookName);
+
+    verify(mockWorkspaceAuthService)
+        .enforceWorkspaceAccessLevel(
+            fromWorkspaceNamespace, fromWorkspaceFirecloudName, WorkspaceAccessLevel.READER);
+    verify(mockWorkspaceAuthService)
+        .enforceWorkspaceAccessLevel(
+            toWorkspaceNamespace, toWorkspaceFirecloudName, WorkspaceAccessLevel.WRITER);
+    verify(mockWorkspaceAuthService)
+        .validateActiveBilling(toWorkspaceNamespace, toWorkspaceFirecloudName);
+
+    FileDetail expectedFileDetail =
+        new FileDetail()
+            .name(newNotebookName + ".ipynb")
+            .path(
+                "gs://"
+                    + TestMockFactory.WORKSPACE_BUCKET_NAME
```

Comment:
Here I'd prefer to see an indication that this is the toWorkspace bucket and not equal to the fromWorkspace bucket.